File: 2020/01/haskell/01.hs
#!/usr/bin/env stack
-- stack --resolver lts-13.7 script

import System.IO
import Data.List

--see https://stackoverflow.com/a/58511843
--combinations of size n
subsequencesOfSize :: Int -> [a] -> [[a]]
subsequencesOfSize n xs = let l = length xs
                          in if (n > l) then []
                             else subsequencesBySize xs !! (l-n)
 where
   subsequencesBySize [] = [[[]]]
   subsequencesBySize (x:xs) = let next = subsequencesBySize xs
                               in zipWith (++)
                                    ([]:next)
                                    ( map (map (x:)) next ++ [[]] )  

day_one :: String -> Int -> Integer -> IO ()
day_one file group goal = do
    --read file by lines
    handle     <- openFile file ReadMode
    contents   <- hGetContents handle
    --type conversion, then get/filter combinations
    let list  = map (read::String->Integer) (lines contents)
        combo = subsequencesOfSize group list
        match = filter (\x -> (sum x) == goal) combo
        prod  = map (\x -> product x) match
    print prod

main = do 
    day_one "../input.txt" 2 2020
    day_one "../input.txt" 3 2020

File: 2020/01/python/01.py
#!/usr/bin/env python 
from itertools import combinations
from functools import reduce
from operator  import mul

def summation(input_list, n, goal):
    #combinations of n numbers that sum to goal
    groups     = combinations(input_list, n)
    select     = filter(lambda x: sum(x) == goal, groups)

    #multiply groups matching goal summation
    answer     = [reduce(mul, x) for x in select]
    
    return list(answer)

if __name__ == "__main__":

    #read input to integer list
    with open("../input.txt", 'r') as f:
        input_list = f.read().splitlines() 
        input_list = list(map(int, input_list))

    p1 = summation(input_list, 
                   n = 2, 
                   goal = 2020)

    p2 = summation(input_list, 
                   n = 3, 
                   goal = 2020)

    print(f"Part 1 answer: {p1}")
    print(f"Part 2 answer: {p2}")
